### PR TITLE
libobs-d3d11: Fixed second source with same display being black

### PR DESCRIPTION
**Pull request by Wouterdek**
This commit fixes a bug that occurs on Windows when 2 or more
"Display Capture" sources with the same display are
simultaneously active.

Only one of these active sources works, all other ones are black.
and error messages with E_INVALIDARG are printed.
The issue can be reproduced by starting OBS, adding 2 display
capture sources with the same display and viewing them both.

The cause of the issue is that a seperate gs_duplicator object is made
for each display capture source. In the constructor for this class,
Start() is called which then calls DuplicateOutput with the display id.
However, the documentation for DuplicateOutput states the following:
  ...
  E_INVALIDARG (is returned) for one of the following reasons:
    The calling application is already duplicating this desktop output.
  ...
source: msdn.microsoft.com/en-us/library/windows/desktop/hh404600.aspx

In conclusion, instances of gs_duplicator must be shared between
display capture sources if the display that is being captured
is the same, which is exactly what this commit implements.

device_duplicator_create now first checks a map to see if an instance
of gs_duplicator already exists for a perticular display id. If not,
a new instance is created and added to the map. In both cases, a
reference counter is incremented and the object is returned.
gs_duplicator_destroy now decrements the reference counter and only
destroys the object if the resulting reference count is zero.
That is, if no more active sources are using this gs_duplicator.

### CODE REVIEW

**Comment by Fenrirthviti** [comment]
Can you explain the use case where Add Existing for a display capture source a second time is not sufficient?

**Comment by WizardCM** [comment]
I'd assume if you wanted different filters on different parts of the screen, like if you use a display capture for one UWP app and a different display capture for half the screen (for example, capturing a VR preview, something I've done before).

**Comment by Wouterdek** [comment]
Also, while Add Existing often works, the current behaviour is clearly a bug. Either users should not be able to create a second "Display Capture" with the same display, OR it should just work. Allowing users to create a second one, but then having it not work without any info is just confusing.

**Comment by dodgepong** [comment]
I'm not sure I would go so far as to say it's a bug (if you try to add a DirectShow device that is already in use, that errors out but is not a bug), but I don't mind smoothing over the user experience for those that don't know about this limitation in Windows.

**Comment by jp9000** [comment]
Woops, unintentionally closed this.

**Comment by jp9000** [comment]
Hey there, welcome to the project, I don't believe I've seen you before.  Good fix, it's along the lines of how I was intending to fix it eventually.  A simple/elegant solution, although there were a few problems with this pull request:

1.) Your commit title and message did not follow our guidelines, and the commit message was much more data than necessary.  It should have been "libobs-d3d11: Allow multiple display captures of same monitor" or something similar, present tense, with a short and concise commit message explaining the commit.  We use the 50/72 style of commit messages.  Please see [the contributing guidelines](https://github.com/jp9000/obs-studio/blob/master/CONTRIBUTING.rst) before making a pull request.

2.) Your code also did not follow our code styling.  I don't quite understand how this happens, I'm guessing you didn't really look at much of the code before modifying it, it's a bit frustrating that some people don't realize that there are other coding styles out there other than the allman code style.  Again please see the contributing guidelines.

3.) You didn't take in to account the fact that it would unnecessarily call IDXGIOutputDuplication::AcquireNextFrame multiple times in a single frame, which would reduce performance unnecessarily and cause additional unnecessary texture copies.  I fixed this so that it checks to see if gs_duplicator_update_frame has been called in the current frame, and if it has been called, it ignores the call and continues.  When a gs_flush has been called, it'll reset the update status of all the current output duplicators.

Still, it was the right direction.  Just in the future when contributing to any open source project, try to be aware of their code and how they operate their project, look at their commit history to get a good gauge on how they make commits and how they code, and read any contribution guidelines so that the maintainer does not have to fix the code or force you to have to rewrite it.  I am probably one of the few that doesn't mind fixing pull requests for the contributors, especially if it's small and I like a contribution enough (and I definitely like this contribution, great work).

I'm merging shortly with the fixes, you'll see "closed in [commit]", that will be the merged commit.

**Comment by Wouterdek** [comment]
Believe it or not, I actually did read the contribution guidelines. I have contributed to a couple of open-source projects before and I fully understand the importance of having and enforcing rules to maintain a consistent style throughout the project.

I checked for trailing spaces, max column width and tabs vs spaces, but I forgot several other things. For future reference, I do not mind cleaning up these mistakes myself. It reduces the workload of project owner and allows me to learn. I'll compare the committed code and my pull request and take a look at the changes.

Thank you for your time and patience.

**Comment by jp9000** [comment]
Hm, well either way it's no problem.  Thank you for this, it was a small and simple fix that probably should have been done a long time ago.
